pytorch_lightning/trainer/distrib_data_parallel.py

```diff
@@ -344,7 +344,7 @@ def ddp_train(self, process_idx, model):
         # copy model to each gpu
         if self.on_gpu:
             self.root_gpu = process_idx
-            self.device = torch.device('cuda', self.root_gpu)
+            self._device = torch.device('cuda', self.root_gpu)
```

Comment:
we could remove all of these calls in the trainer by overloading `.to()` and `.cuda()` in LightningModule and setting the device there.

Reply:
You also need to overload `.cpu()` :)

Reply:
sorry, maybe i’m missiny something. The point of self.device  is to have a readonly property to create tensors in memory directly.

Reply:
If we overload the the `.to()` method like this for example:
```python
def to(device):
    super().to(device)
    self._device = device
```
Then we get the following benefits:

- self.device property will not break when LightningModule is used as nn.Module without Trainer
- When LightningModule is a nested LightningModule and user calls .to(), also the self.device properties of submodules get updated
- The Trainer code does not need to set the device, it calls .to anyway, so the code is in one place and is easier to maintain.

I see only benefits atm

Reply:
@justusschock also did it like this for metrics package

Reply:
oh i see. cool!

sounds reasonable

Reply:
so we need to overwrite the following methods:
- `.to(...)`
- `.cpu()`
- `.cuda()`
or am I missing any? @awaelchli ^^

Reply:
yep, exactly, although I  suspect cpu and cuda already call .to internally. Not sure, need to check.

Reply:
well it seems to me that ideally, we want to rise the whole template from metrics...

Reply:
probably not all. device.setter, half, double, dtype all of that stuff does not apply for LightningModule. I agree we should try to avoid code duplication.
@justusschock what do you think?

Reply:
I think, while we do this, we should think about introducing the same for dtype, since when I create a tensor in a function, it usually involves a certain dtype as well. Although I'm not sure, if this would be reflected by amp as well...